source/1_engine/rex_engine/include/rex_engine/diagnostics/logging/internal/sinks/unix/ansicolor_sink.h

```diff
@@ -0,0 +1,238 @@
+
+      // Foreground colors
+      const string_view_t black   = "\033[30m";
+      const string_view_t red     = "\033[31m";
+      const string_view_t green   = "\033[32m";
+      const string_view_t yellow  = "\033[33m";
+      const string_view_t blue    = "\033[34m";
+      const string_view_t magenta = "\033[35m";
+      const string_view_t cyan    = "\033[36m";
+      const string_view_t white   = "\033[37m";
+
+      /// Background colors
+      const string_view_t on_black   = "\033[40m";
+      const string_view_t on_red     = "\033[41m";
+      const string_view_t on_green   = "\033[42m";
+      const string_view_t on_yellow  = "\033[43m";
+      const string_view_t on_blue    = "\033[44m";
+      const string_view_t on_magenta = "\033[45m";
+      const string_view_t on_cyan    = "\033[46m";
+      const string_view_t on_white   = "\033[47m";
+
+      /// Bold colors
+      const string_view_t yellow_bold = "\033[33m\033[1m";
+      const string_view_t red_bold    = "\033[31m\033[1m";
+      const string_view_t bold_on_red = "\033[1m\033[41m";
+
+    private:
+      FILE* m_target_file;
+      mutex_t& m_mutex;
+      bool should_do_colors_;
+      PatternFormatter m_formatter;
+      rsl::array<rex::DebugString, rsl::enum_refl::enum_count<level::LevelEnum>()> m_colors;
+      void print_ccode_impl(const string_view_t& color_code);
+      void print_range_impl(const memory_buf_t& formatted, size_t start, size_t end);
+      static rex::DebugString to_string_impl(const string_view_t& sv);
+    };
+
+    template <typename ConsoleMutex>
+    ansicolor_sink<ConsoleMutex>::ansicolor_sink(FILE* target_file, ColorMode mode)
+        : m_target_file(target_file)
+        , m_mutex(ConsoleMutex::mutex())
+        , m_formatter(details::make_unique<rexlog::pattern_formatter>())
+
+    {
+      set_color_mode(mode);
+      m_colors[level::trace]    = to_string_impl(white);
+      m_colors[level::debug]    = to_string_impl(cyan);
+      m_colors[level::info]     = to_string_impl(green);
+      m_colors[level::warn]     = to_string_impl(yellow_bold);
+      m_colors[level::err]      = to_string_impl(red_bold);
+      m_colors[level::critical] = to_string_impl(bold_on_red);
+      m_colors[level::off]      = to_string_impl(reset);
+    }
+
+    template <typename ConsoleMutex>
+    void ansicolor_sink<ConsoleMutex>::set_color(level::LevelEnum color_level, string_view_t color)
+    {
+      rsl::lock_guard<mutex_t> lock(m_mutex);
+      m_colors[static_cast<size_t>(color_level)] = to_string_impl(color);
+    }
+
+    template <typename ConsoleMutex>
+    void ansicolor_sink<ConsoleMutex>::log(const details::LogMsg& msg)
+    {
+      // Wrap the originally formatted message in color codes.
+      // If color is not supported in the terminal, log as is instead.
+      rsl::lock_guard<mutex_t> lock(m_mutex);
+      msg.color_range_start = 0;
+      msg.color_range_end   = 0;
+      memory_buf_t formatted;
```

Comment:
I think it's best to use a stack string here instead of regular string

Reply:
I changed it some place else as well, we must be aware though as this is the string for the log message itself I made it a big_stack_string but this will limit the message to 256 characters. I hope this is enough for all messages.

Reply:
that should be fine, if you're concerned, the string returned by rsl::format is 512 characters max, these strings only live for a very short time as they're often copied elsewhere or stored into a smaller destination string so it's not an issue (if it is, it'll be caught by the assert in stack string)